### PR TITLE
[Misc] M2m 0724 continuation

### PR DESCRIPTION
**Pull request by czc-unac**
### What this PR does / why we need it?

Adapt vllm-ascend to vLLM main commits up to July 24.

### Changes

| Files | Upstream vLLM Change | vllm-ascend Adaptation |
|---|---|---|
| `vllm_ascend/worker/v2/attn_utils.py`<br>`vllm_ascend/worker/v2/spec_decode/dflash/speculator.py`<br>`vllm_ascend/worker/v2/spec_decode/dspark/speculator.py` | [vllm#44492](https://github.com/vllm-project/vllm/pull/44492) ([4ec199b6](https://github.com/vllm-project/vllm/commit/4ec199b6)) — Populate draft `seq_lens_cpu_upper_bound` for spec-decode attention metadata. Upstream threads `seq_lens_cpu_upper_bound` through `_multi_step_decode()` (new 5th positional arg) and `_build_draft_attn_metadata()` (new `seq_lens_cpu_upper_bound` + `step` kwargs). | `attn_utils.py`: Added `build_attn_metadata` with `seq_lens_cpu_upper_bound` param (defaults to `seq_lens_cpu`) and `build_attn_metadata_wrapper()` context manager for Ascend NPU attention metadata building.<br>`dflash/speculator.py`: `propose()` caches `self.input_batch` before `super().propose()`; `build_draft_attn_metadatas()` passes actual `num_reqs`, padded counts, `seq_lens_cpu_upper_bound`, and `step` to `_build_draft_attn_metadata`.<br>`dspark/speculator.py`: `build_draft_attn_metadatas()` passes `seq_lens_cpu_upper_bound` + `step`; `propose()` caches `self.input_batch` (version-gated for main only, since v0.25.1 never reaches this code path via `run_fullgraph`). |
| `vllm_ascend/worker/v2/spec_decode/dflash/aclgraph.py` | [vllm#47914](https://github.com/vllm-project/vllm/pull/47914) ([0d12618e](https://github.com/vllm-project/vllm/commit/0d12618e)) — `causal` moved from `DFlashCudaGraphManager.__init__` to `capture()`. | `capture()` is version-gated: passes `causal` kwarg only on main branch; v0.25.1 receives it via `__init__`. `run_fullgraph()` accesses `self.speculator.input_batch.seq_lens_cpu_upper_bound` for building draft attention metadata during full-graph replay, then updates Ascend NPU full-graph params. |
| `vllm_ascend/core/recompute_scheduler.py` | [vllm#47312](https://github.com/vllm-project/vllm/pull/47312) ([12213c67](https://github.com/vllm-project/vllm/commit/12213c67)) — Handle grammar compilation failures. `finish_requests()` return type changed from `list[tuple(str, int)]` to `list[Request]`. | Assertion in `_finish_recomputed_request` is version-gated: on main branch extracts `(request_id, client_index)` from `Request` objects; v0.25.1 retains the direct tuple comparison. |
| `vllm_ascend/_310p/model_runner_310p.py` | General vLLM V1 API evolution (model loading, KV cache allocation, `determine_available_memory` flow). | New file. 310P hardware platform model runner extending `NPUModelRunner`. Adapts 310P-specific ACL graph capture/replay, splitfuse attention, NPU input batch, and FRACTAL_NZ weight layout to the upstream V1 API. |
| `.github/vllm-main-verified.commit` | N/A (CI tracking) | Updated to `d02df748bf9efd99022f1a062597dc3cb3808485` to record the verified upstream vLLM commit. |

### How was this patch tested?

- vLLM version: v0.25.1
- vLLM main: https://github.com/vllm-project/vllm/commit/fe784ff22e630a31fd798f392b01e0a75c18f047

### CODE REVIEW

**Comment by zhangxinyuehfad** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Comment by czc-unac** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by czc-unac** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by czc-unac** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by czc-unac** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Comment by czc-unac** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Comment by czc-unac** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

File: vllm_ascend/worker/v2/spec_decode/dflash/speculator.py [hunk 1]
```diff
@@ -81,9 +81,11 @@ def build_draft_attn_metadatas(self, num_reqs_padded):
         num_tokens_padded = num_reqs_padded * self.num_query_per_req
         with build_attn_metadata_wrapper():
             attn_metadata = self._build_draft_attn_metadata(
-                num_reqs=num_reqs_padded,
+                num_reqs=self.input_batch.num_reqs,
```
Comment (wxsIcey) [inline vllm_ascend/worker/v2/spec_decode/dflash/speculator.py]: Why the change here? This can cause tensor shape mismatch during aclgraph replay.

File: vllm_ascend/worker/v2/spec_decode/dspark/speculator.py [hunk 2]
```diff
@@ -86,9 +86,11 @@ def build_draft_attn_metadatas(self, num_reqs_padded):
         num_tokens_padded = num_reqs_padded * self.num_query_per_req
         with build_attn_metadata_wrapper():
             attn_metadata = self._build_draft_attn_metadata(
-                num_reqs=num_reqs_padded,
+                num_reqs=self.input_batch.num_reqs,
```
Comment (wxsIcey) [inline vllm_ascend/worker/v2/spec_decode/dspark/speculator.py]: ditto

File: vllm_ascend/worker/v2/spec_decode/dflash/speculator.py (hunk 1 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/dflash/speculator.py]: Sure, I will change it;
Pad seq_lens_cpu_upper_bound to num_reqs_padded, keep num_reqs=num_reqs_padded.
```python
upper_bound = self.input_batch.seq_lens_cpu_upper_bound
if upper_bound.shape[0] < num_reqs_padded:
    upper_bound= torch.nn.functional.pad(upper_bound, (0, num_reqs_padded - ub.shape[0]))
```
- num_reqs unchanged → query_start_loc_cpu stays monotonic in both capture and replay → actual_seq_lengths_q always [1,1,1,1,...] → zero FIA replay risk
- upper_bound padded → seq_lens_cpu_upper_bound[:num_reqs] never OOB
- Capture path untouched (ub.shape[0] == num_reqs_padded, branch skipped)

File: vllm_ascend/worker/v2/spec_decode/dspark/speculator.py [hunk 3]
```diff
@@ -86,9 +86,11 @@ def build_draft_attn_metadatas(self, num_reqs_padded):
         num_tokens_padded = num_reqs_padded * self.num_query_per_req
         with build_attn_metadata_wrapper():
             attn_metadata = self._build_draft_attn_metadata(
```
Comment (zhangxinyuehfad) [inline vllm_ascend/worker/v2/spec_decode/dspark/speculator.py:94]: I have adapted these parameters in `_build_draft_attn_metadata`.

**Comment by zhangxinyuehfad** [comment]
e2e full test ok: https://github.com/vllm-project/vllm-ascend/actions/runs/30235923100/job/89899871860?pr=12648

File: vllm_ascend/worker/v2/spec_decode/dflash/speculator.py [hunk 4]
```diff
@@ -79,11 +79,16 @@ def set_attn(
     # keep the current name for now as upstream may change it again.
     def build_draft_attn_metadatas(self, num_reqs_padded):
         num_tokens_padded = num_reqs_padded * self.num_query_per_req
+        upper_bound = self.input_batch.seq_lens_cpu_upper_bound
+        if upper_bound.shape[0] < num_reqs_padded:
+            upper_bound = torch.nn.functional.pad(upper_bound, (0, num_reqs_padded - upper_bound.shape[0]))
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/spec_decode/dflash/speculator.py]: this is overcooked, let's just pass `self.input_batch.seq_lens_cpu_upper_bound` by.

File: vllm_ascend/worker/v2/spec_decode/dflash/speculator.py [hunk 5]
```diff
@@ -79,11 +79,16 @@ def set_attn(
...
         with build_attn_metadata_wrapper():
             attn_metadata = self._build_draft_attn_metadata(
                 num_reqs=num_reqs_padded,
                 num_reqs_padded=num_reqs_padded,
                 num_tokens_padded=num_tokens_padded,
+                seq_lens_cpu_upper_bound=upper_bound,
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/spec_decode/dflash/speculator.py]:
```suggestion
                seq_lens_cpu_upper_bound=self.input_batch.seq_lens_cpu_upper_bound,
```

File: vllm_ascend/worker/v2/spec_decode/dspark/speculator.py [hunk 6]
```diff
@@ -84,11 +84,16 @@ def set_attn(
 
     def build_draft_attn_metadatas(self, num_reqs_padded):
         num_tokens_padded = num_reqs_padded * self.num_query_per_req
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/spec_decode/dspark/speculator.py:92]: let's add a new input arg `seq_lens_cpu_upper_bound` here instead of adding a new instance member `self.input_batch` for class `AscendDSparkSpeculator`

File: vllm_ascend/worker/v2/spec_decode/dspark/speculator.py [hunk 7]
```diff
@@ -84,11 +84,16 @@ def set_attn(
...
+        upper_bound = self.input_batch.seq_lens_cpu_upper_bound
+        if upper_bound.shape[0] < num_reqs_padded:
+            upper_bound = torch.nn.functional.pad(upper_bound, (0, num_reqs_padded - upper_bound.shape[0]))
         with build_attn_metadata_wrapper():
             attn_metadata = self._build_draft_attn_metadata(
                 num_reqs=num_reqs_padded,
                 num_reqs_padded=num_reqs_padded,
                 num_tokens_padded=num_tokens_padded,
+                seq_lens_cpu_upper_bound=upper_bound,
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/spec_decode/dspark/speculator.py]: ditto.

File: vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py [hunk 8]
```diff
@@ -101,16 +102,27 @@ def create_forward_fn(desc: BatchExecutionDescriptor, warmup: bool):
...
+                if vllm_version_is("0.25.1"):
+                    prepare_inputs_to_capture(
+                        num_reqs,
+                        num_tokens,
+                        model_state,
+                        input_buffers,
+                        block_tables,
+                        attn_groups,
+                        kv_cache_config,
+                        full_cudagraph=desc.cg_mode == CUDAGraphMode.FULL,
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py]: why do we change this?

File: vllm_ascend/worker/v2/spec_decode/eagle/speculator.py [hunk 9]
```diff
@@ -276,20 +277,28 @@ def _multi_step_decode(
             assert self.decode_cudagraph_manager is not None
             self.decode_cudagraph_manager.run_fullgraph(batch_desc)
             return
-        super()._multi_step_decode(num_reqs, skip_attn, batch_desc, num_tokens_across_dp)
+        # The cudagraph-capture forward_fn lambda does not pass
+        # seq_lens_cpu_upper_bound; fall back to the draft input buffers.
+        if seq_lens_cpu_upper_bound is None:
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/spec_decode/eagle/speculator.py]: let's use vllm version check, when using tha latest commit id, `assert seq_lens_cpu_upper_bound is not None`.

File: vllm_ascend/worker/v2/attn_utils.py [hunk 10]
```diff
@@ -127,6 +128,8 @@ def build_attn_metadata(
     if seq_lens_np is None:
         seq_lens_np = np.full(num_reqs, max_seq_len, dtype=np.int32)
     seq_lens_cpu = torch.from_numpy(seq_lens_np)[:num_reqs]
+    if seq_lens_cpu_upper_bound is None:
+        seq_lens_cpu_upper_bound = seq_lens_cpu
```
Comment (MengqingCao) [inline vllm_ascend/worker/v2/attn_utils.py]: ditto.

File: vllm_ascend/core/recompute_scheduler.py [hunk 11]
```diff
@@ -143,7 +143,12 @@ def _finish_recomputed_request(
             request.request_id,
             RequestStatus.FINISHED_ABORTED,
         )
-        assert finished_reqs == [(request.request_id, request.client_index)]
+        if vllm_version_is("0.25.1"):
```
Comment (MengqingCao) [inline vllm_ascend/core/recompute_scheduler.py:146]: @nwpu-zxr plz review the changes on recompute scheduler.

File: vllm_ascend/worker/v2/attn_utils.py (hunk 10 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/attn_utils.py]: done

File: vllm_ascend/worker/v2/spec_decode/eagle/speculator.py (hunk 9 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/eagle/speculator.py]: Done

File: vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py (hunk 8 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py]: This is an unnecessary modification, I will revert。

File: vllm_ascend/worker/v2/spec_decode/dspark/speculator.py (hunk 7 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/dspark/speculator.py]: done

File: vllm_ascend/worker/v2/spec_decode/dspark/speculator.py (hunk 6 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/dspark/speculator.py:92]: It's a better idea, I will change it.

File: vllm_ascend/worker/v2/spec_decode/dflash/speculator.py (hunk 4 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/dflash/speculator.py]: Sure, done

File: vllm_ascend/core/recompute_scheduler.py (hunk 11 above)
Comment (nwpu-zxr) [reply vllm_ascend/core/recompute_scheduler.py:146]: Approve.

File: vllm_ascend/worker/v2/spec_decode/dflash/aclgraph.py [hunk 12]
```diff
@@ -105,7 +105,10 @@ def run_fullgraph(self, desc: BatchExecutionDescriptor) -> torch.Tensor | tuple[
         """Override run_fullgraph to update full graph params in run_fullgraph."""
         num_tokens = desc.num_tokens
 
-        draft_attn_metadatas = self.speculator.build_draft_attn_metadatas(desc.num_reqs)
+        draft_attn_metadatas = self.speculator.build_draft_attn_metadatas(
```
Comment (wxsIcey) [inline vllm_ascend/worker/v2/spec_decode/dflash/aclgraph.py:108]: shoud we add the version judge?

```
if vllm_version_is("0.25.1"):
    draft_attn_metadatas = self.speculator.build_draft_attn_metadatas(desc.num_reqs)
else:
     draft_attn_metadatas = self.speculator.build_draft_attn_metadatas(desc.num_reqs, self.speculator.input_batch.seq_lens_cpu_upper_bound)

```

File: vllm_ascend/worker/v2/spec_decode/dflash/speculator.py [hunk 13]
```diff
@@ -77,13 +78,15 @@ def set_attn(
 
     # NOTE: upstream vLLM named this to _build_draft_attn_metadatas;
     # keep the current name for now as upstream may change it again.
-    def build_draft_attn_metadatas(self, num_reqs_padded):
+    def build_draft_attn_metadatas(self, num_reqs_padded, seq_lens_cpu_upper_bound):
```
Comment (wxsIcey) [inline vllm_ascend/worker/v2/spec_decode/dflash/speculator.py:81]: ditto, if we don‘t support mrv2 in specical version, please remove all judge.

Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/dflash/speculator.py:81]: We will drop it when upgrade 0.26.0

File: vllm_ascend/worker/v2/spec_decode/dflash/aclgraph.py (hunk 12 above)
Comment (zhangxinyuehfad) [reply vllm_ascend/worker/v2/spec_decode/dflash/aclgraph.py:108]: We will drop it when upgrade 0.26.0

**Review by wxsIcey — APPROVED** [review]
Change to the spec decode in mrv2 is right. 

Todo: Please remove the version jude in mrv2.

**Review by wangxiyuan — APPROVED** [review]
